feat(pipelines): add ecr source action

alexpulver opened the pull request:
Closes #16378 

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 641664c94aa8449fdac8370e46f06829b0dc1255
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d848713f-01e0-4204-9b1c-8da4849587b1%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

alexpulver commented:
I will need to rework the `actionName` definition, because `/` can appear in ECR repository name (namespace), and is not a valid character in CodePipeline action name.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 0bfd98cd903aadc8c52d35f9f0cde43efbcf8a7e
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=c2ecf2f2-a696-4cef-8953-bd3f3b035450%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: e2e124c029af5c73bc584e930f66224b3718f9f5
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=03ab0b6d-133b-43ca-ab86-e4d21a8218bd%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

alexpulver commented:
@rix0rrr I changed the code to handle the `/` namespace character by replacing it with `_` character. I tested manually that `const formattedRepositoryName = Fn.join('_', Fn.split('/', this.repository.repositoryName));` works on repositories with and without namespace, because I don't think it's possible to test without deploying the stack. Below are the snippets from the template I used for manual testing. Can you please let me know if it's acceptable from your perspective?

No namespace:
```json
{
  "Resources": {
    "Repository22E53BBD": {
      "Type": "AWS::ECR::Repository"
    },
<snip>
```

With namespace:
```json
{
  "Resources": {
    "Repository22E53BBD": {
      "Type": "AWS::ECR::Repository",
      "Properties": {
        "RepositoryName": "namespace/repo"
      }
    },
<snip>
```

Pipeline:
```json
{
  "Resources": {
    <snip>
    "Pipeline9850B417": {
      "Type": "AWS::CodePipeline::Pipeline",
      "Properties": {
        "RoleArn": {
          "Fn::GetAtt": [
            "PipelineRoleB27FAA37",
            "Arn"
          ]
        },
        "Stages": [
          {
            "Actions": [
              {
                "InputArtifacts": [],
                "ActionTypeId": {
                  "Version": "1",
                  "Owner": "AWS",
                  "Category": "Source",
                  "Provider": "ECR"
                },
                "OutputArtifacts": [
                  {
                    "Name": "Source"
                  }
                ],
                "RunOrder": 1,
                "Configuration": {
                  "ImageTag": "latest",
                  "RepositoryName": {
                    "Ref": "Repository22E53BBD"
                  }
                },
                "Name": {
                  "Fn::Join": [
                    "_",
                    {
                      "Fn::Split": [
                        "/",
                        {
                          "Ref": "Repository22E53BBD"
                        }
                      ]
                    }
                  ]
                }
              }
            ],
            "Name": "Source"
          },
<snip>
```

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 910ef81282c220753a1efd0167b52a73ba445478
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=e41aedfe-026a-40e3-8641-6d454a127e0e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

alexpulver commented:
@rix0rrr would be glad to have your input on the above

npvisual commented on packages/@aws-cdk/pipelines/lib/codepipeline/codepipeline-source.ts:
```diff
@@ -51,16 +52,37 @@ export abstract class CodePipelineSource extends Step implements ICodePipelineAc
-   * const bucket: IBucket = ...
-   * CodePipelineSource.s3(bucket, {
-   *   key: 'path/to/file.zip',
+   * import * as ecr from '@aws-cdk/aws-ecr';
```
@alexpulver  : should the following comments have been modified ? I think they still apply to the `s3` factory method.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 7d82beee845afe1992c1475ff882bf90c6fd4c26
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d52035ba-465a-4565-890b-9ca18d6a5ebc%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 970c27324c4ce4b2332aad9cae7b9ae63925a94d
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=5c3a9b0c-708d-40f6-9129-3ebb24e5f0b8%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: aa2a00e45487e86968128863e0ceea163196f828
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=757db201-d95b-44f0-a6b0-14101f1b01f3%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 1761505ffde188b9c20358ec8128b92fc07e06ac
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=cd379f88-bba4-46b8-8f4a-bf00706629a5%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
